text/0001-remove_ambiguous_grammar.md

```diff
@@ -0,0 +1,234 @@
+
+```sql
+DELETE some_data FROM values WHERE if CONTAINS true
+```
+
+The above query checks for deletion if the identifier `if` contains true. It is not parsed as an `IF` statement since the parser fails when trying to parse a code block when it finds the `true` token, and therefore falls back to parsing a the production as a value.
+
+Now suppose the user wanted to make a change and add a expression block:
+
+```sql
+DELETE some_data FROM values WHERE if CONTAINS { true }
+```
+
+Now the statement is suddenly parsed as an IF statement and, if the table happens to have a `CONTAINS` field, it will still run but with a completely different meaning.
+
+A different type of ambiguity in today's grammar is token ambiguity.
+Most languages try to limit the possible meaning of tokens as much as possible.
+As apposed to most languages in SurrealQL raw Identifiers can start with a digit. This means that a number is a valid identifier, and a number with a fraction could also be parsed as a identifier and its field. This can make it hard to determine what is an identifier. Consider the following example:
+
+```sql
+UPDATE table SET 10dec = "a value" WHERE 10dec = "a different value";
+```
+
+In the above query the field `10dec` will be updated to the value "a value" if the number 10 is equal to "a different value". In SET you are not allowed to have any numbers so `10dec` which is a valid identifier will be parsed as an identifier. In the where clause however numbers are allowed so `10dec` will in that position be parsed as number.
+
+More examples of these types of confusing grammar can probably be found.
+
+## Hard to specify
+
+As we are working towards a production ready database we should probably eventually create an official specification for the SurrealQL language. Most languages are either Context-Free or almost Context-Free and therefore allow specifying the grammar in Context-Free grammar. See for example the JavaScript spec: [https://tc39.es/ecma262/#prod-Statement](https://tc39.es/ecma262/#prod-Statement)
+
+The current ambiguities in the language make it impossible to specify a grammar for SurrealQL in the same way.
+
+## Extendibility
+
+Because a lot of the current syntax is defined in part by the failure to parse other productions, productions which today are parsed one why could in the future silently change to have a completely different semantic meaning.
+
+Consider the following query:
+
+```sql
+SELECT count FROM table GROUP count
+```
+
+Currently the values from this select will be grouped by the field key. But where we to every introduce some feature which would require the `COUNT` keyword after `GROUP` then this code will quietly change meaning.
+
+# 3. Proposal
+
+## Location specific reserved words.
+
+The first change proposed is to disallow raw identifiers in places where the same identifier could also be parsed as a keyword. For example, the following code would no longer be allowed:
```

Comment:
Could it be more simple to just disallow certain keywords, and not allowing for it be be wrapped? Would users be very upset by being unable to use a few keywords?